sdks/python/apache_beam/runners/worker/operations.py

```diff
@@ -216,6 +216,9 @@ def finish(self):
     """Finish operation."""
     pass
 
+  def teardown(self):
```

Comment:
Can we avoid adding setup() and teardown() to the general Operation class?  I don't expect operations other than DoOperation to have these lifecycle functions.  Instead, we should call DoFnRunner.setup() in DoOperation.start() and DoFnRunner.teardown() in DoOperation.finish().  This way, we don't need to pipe changes up into bundle_processor.py and fn_api_runner.py.

Reply:
Ah, I did not realize that  `Operation.start` and `Operation.finish` was called once per instance rather than once per bundle.

The main reason for moving `setup` and `teardown` to `Operation` would be to support adding `setup` and `teardown` to `CombineFn` for [BEAM-3736](https://issues.apache.org/jira/browse/BEAM-3736). That said, the change can be made directly in `PGBKCVOperation` and `FlattenOperation` instead.

Reply:
Offline discussion: for most runners, the DoOperation is instantiated once per bundle, rather than being long-lived between bundles. Therefore there are no performance benefits to this change. However, we should still add these methods to the API so that if / when the implementation is changed to support long-lived DoOperation, users will automatically gain those benefits.

Reply:
Most, and hopefully in the future all, runners should be executing via the FnAPI which do cache these operation trees, so it does make sense to put this into operation and lift it up to a method on bundle_processor.py

Reply:
@charlesccychen what's your opinion on this?

Reply:
Adding the new lifecycle methods to the base class makes sense to me.

Do you mind adding pydoc comments to start/finish/setup/teardown etc., such that the next person to make a change here can understand what is being called when.

Reply:
Added comments.